dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/CommonMergeHandler.java

```diff
@@ -0,0 +1,107 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.merge;
+
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.hisp.dhis.common.BaseIdentifiableObject;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.common.collection.CollectionUtils;
+import org.hisp.dhis.dataentryform.DataEntryForm;
+import org.hisp.dhis.dataentryform.DataEntryFormService;
+import org.hisp.dhis.indicator.Indicator;
+import org.hisp.dhis.indicator.IndicatorService;
+import org.springframework.stereotype.Service;
+
+/**
+ * Common Merge handler for metadata entities. The merge operations here are shared by many merge
+ * use cases (e.g. Indicator merge & DataElement merge), hence the need for common handlers, to
+ * reuse code and avoid duplication.
+ *
+ * @author david mackessy
+ */
+@Service
+@RequiredArgsConstructor
+public class CommonMergeHandler {
+
+  private final IndicatorService indicatorService;
+  private final DataEntryFormService dataEntryFormService;
+
+  /**
+   * Replace all {@link BaseIdentifiableObject} refs in the numerator and denominator fields of
+   * {@link Indicator}s
+   *
+   * @param sources to be replaced
+   * @param target to replace each source
+   */
+  public <T extends BaseIdentifiableObject> void handleRefsInIndicatorExpression(
+      List<T> sources, T target) {
+    for (T source : sources) {
+      // numerators
+      List<Indicator> numeratorIndicators =
+          indicatorService.getIndicatorsWithNumeratorContaining(UID.of(source.getUid()));
+
+      numeratorIndicators.forEach(
+          i -> {
+            String existingNumerator = i.getNumerator();
+            i.setNumerator(existingNumerator.replace(source.getUid(), target.getUid()));
+          });
+
+      // denominators
+      List<Indicator> denominatorIndicators =
+          indicatorService.getIndicatorsWithDenominatorContaining(UID.of(source.getUid()));
+
+      denominatorIndicators.forEach(
+          i -> {
+            String existingDenominator = i.getDenominator();
+            i.setDenominator(existingDenominator.replace(source.getUid(), target.getUid()));
+          });
```

Comment:
Yep, I thought about this. I went with the more explicit and targeted approach (1 for each). It's debatable which would be better.
Best case scenario is that the Indicators have the same DE refs in both fields.
Worst case scenario is they don't. It would depend how the implementation has set up their system. Seems like it comes down to deciding whether we want:
(i) less IO (1 DB call) with potentially more in-memory code execution vs 
(ii) more IO (2 DB calls) with less in-memory code execution

happy to change it to 1 DB call. Only testing with large quantities of data would show the more performant approach. WDYT?

Reply:
I was just thinking doing it together is less hassle. 1 query, one update per indicator. Not because of performance but simplicity of operation. But this isn't very important, just a thought and down to personal preference.